### ISSUE TITLE
`getAccessibleName` fails with `Maximum call stack size exceeded`

### ISSUE THREAD

**Issue by strmer15** [eps1lon/dom-accessibility-api#897]
In version 0.5.15 of `dom-accessibility-api`, our tests started failing with `RangeError: Maximum call stack size exceeded`. Previously, we had been using version 0.5.14, but I had updated our lock file to get the latest transitive dependency versions.

The full stack trace is
```
RangeError: Maximum call stack size exceeded
    at CSSStyleDeclaration.get [as cssText] (https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:30:82799)
    at CSSStyleDeclaration._setProperty (https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:30:82082)
    at CSSStyleDeclaration.set (https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:30:140717)
    at CSSStyleDeclaration.<anonymous> (https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:30:112985)
    at Array.forEach (<anonymous>)
    at CSSStyleDeclaration.<anonymous> (https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:30:112910)
    at CSSStyleDeclaration.set [as font] (https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:30:145022)
    at CSSStyleDeclaration.setProperty (https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:30:81653)
    at https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:37:49908
    at CSSStyleDeclaration.forEach (<anonymous>)
    at https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:37:49869
    at https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:37:1620704
    at Array.forEach (<anonymous>)
    at handleSheet (https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:37:1620484)
    at exports.forEachMatchingSheetRuleOfElement (https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:37:1620808)
    at Window.getComputedStyle (https://59jbmi.csb.app/static/js/jsdom-16.3.0.min.js:37:49816)
    at isHidden (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:22:15)
    at computeTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:275:21)
    at eval (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:143:20)
    at Array.forEach (<anonymous>)
    at computeMiscTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:142:16)
    at computeTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:339:31)
    at eval (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:282:16)
    at Array.map (<anonymous>)
    at computeTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:281:28)
    at eval (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:143:20)
    at Array.forEach (<anonymous>)
    at computeMiscTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:142:16)
    at computeTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:339:31)
    at eval (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:282:16)
    at Array.map (<anonymous>)
    at computeTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:281:28)
    at eval (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:143:20)
    at Array.forEach (<anonymous>)
    at computeMiscTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:142:16)
    at computeTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:339:31)
    at eval (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:282:16)
    at Array.map (<anonymous>)
    at computeTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:281:28)
    at eval (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:143:20)
    at Array.forEach (<anonymous>)
    at computeMiscTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:142:16)
    at computeTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:339:31)
    at eval (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:282:16)
    at Array.map (<anonymous>)
    at computeTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:281:28)
    at eval (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:143:20)
    at Array.forEach (<anonymous>)
    at computeMiscTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:142:16)
    at computeTextAlternative (https://59jbmi.csb.app/node_modules/dom-accessibility-api/dist/accessible-name-and-description.mjs:339:31)
```

See a reproduction of this issue at https://codesandbox.io/s/typescript-jest-test-template-forked-59jbmi?file=/src/index.ts
If you change the version back to 0.5.14, save and then refresh, you'll see that the test passes.

**Comment by strmer15** [comment]
Just FYI that I also created this Jest test on my local dev machine:

```
import { computeAccessibleName } from 'dom-accessibility-api';

import * as util from 'util';

describe('accessibility-api', () => {
  beforeAll(() => {
    Object.defineProperty(globalThis, 'TextEncoder', {
      writable: true,
      value: util.TextEncoder,
    });
    Object.defineProperty(globalThis, 'TextDecoder', {
      writable: true,
      value: util.TextDecoder,
    });
  });

  test('it gets the accessible name', async () => {
    const html = `
      <div id="root">
        <label for="toggle-button" id="label" value="full">
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="label toggle-button"
            id="toggle-button"
            type="button"
          >
            <span>
              Full Refund
            </span>
          </button>
          <div>
            Refund Type
          </div>
        </label>
      </div>`;

    const dom = new (await import('jsdom')).JSDOM(html);

    const element = dom.window.document.getElementById('toggle-button');
    if (element === null) {
      throw new Error('Element is null');
    }

    const name = computeAccessibleName(element);

    expect(name).toBe('Refund Type');
  });
});
```

This is with Node 18.12.1, Jest 29.3.1, and JSDOM 20.0.3.

**Comment by mattmikolay** [comment]
I’m also encountering an uncaught `RangeError: Maximum call stack size exceeded` in a similar scenario, when attempting to call `computeAccessibleName` on a DOM hierarchy that looks like this:

```html
<h2 id="lorem-heading">
  Lorem ipsum
  <a aria-labelledby="lorem-heading" href="#lorem-heading" tabindex="-1">
    <svg>…</svg>
  </a>
</h2>
```

From what I can tell, the issue occurs because calling `computeTextAlternative` for the inner `<a>` element will trigger infinite recursion. `computeTextAlternative` will follow the `aria-labelledby` attribute to the `<h2>`, and then `computeMiscTextAlternative` will attempt to compute the heading’s accessible name by calling `computeTextAlternative` on all of the `<h2>` element’s children. The original `<a>` element itself is one of these children, which causes the recursion.

I’m not super familiar with the W3C spec, but I noticed the following paragraph for step 2F:

> Each node in the subtree is consulted only once. If text has been collected from a descendant, but is referenced by another IDREF in some descendant node, then that second, or subsequent, reference is not followed. This is done to avoid infinite loops.

Maybe I’m misunderstanding, but doesn’t this mean the `computeMiscTextAlternative` function should filter `childNodes` to remove `node`?

**Comment by mattmikolay** [comment]
Thanks for the quick fix, @eps1lon! 🙏
